### PR TITLE
Reerror: error-resilient parsing (same as Merlin for OCaml)

### PR DESCRIPTION
**Pull request by let-def**
This PR implements a new "parsing loop" that is more robust to errors and should improve the experience in an IDE.

The parser itself is not affected (we still have a declarative grammar), but we now have a slightly different interpretation in interactive use-cases.

### CODE REVIEW

**Comment by let-def** [comment]
Esy is missing the "fix" dependency. I don't know the proper way to update it.

**Comment by ulrikstrid** [comment]
I can add `@opam/fix` via a resolution in a package.json. I wonder if this is a bug with parsing the .opam file in `esy`.

cc @andreypopp

**Comment by jordwalke** [comment]
Been testing this out and it's so good.
@ulrikstrid - it wouldn't be a resolution, it would just be a regular dependency.

@let-def you just add `"@opam/fix": "*"` to the "dependencies" section of `src/refmt/esy.json` and `./esy.json` (the later is just the development-time package configuration, the later is what will be released).

If you just add those fields, the CI should pass and we can release this puppy.

File: src/reason-parser/TODO [hunk 1]
```diff
@@ -0,0 +1,18 @@
+Recovery (Urgent):
```
Comment (jordwalke) [inline src/reason-parser/TODO]: Did you want to delete this file and the Plan document?

File: src/menhir-recover/Makefile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+.PHONY: all clean
```
Comment (jordwalke) [inline src/menhir-recover/Makefile]: Was this (menhir-recover) copied form another project somewhere?

File: src/reason-parser/TODO (hunk 1 above)
Comment (let-def) [reply src/reason-parser/TODO]: Yes

File: src/menhir-recover/Makefile (hunk 2 above)
Comment (let-def) [reply src/menhir-recover/Makefile]: Yes, it is a slightly improved version of https://github.com/ocaml/merlin/tree/master/src/ocaml/preprocess/recover (the same algorithm I wrote for Merlin/OCaml parser).

Comment (jordwalke) [reply src/menhir-recover/Makefile]: Would you upstream your changes into merlin?

Comment (let-def) [reply src/menhir-recover/Makefile]: I still have some work to do to generalize the algorithm a bit, but I plan to upstream it either into a separate repository or into Menhir directly.

**Comment by jordwalke** [comment]
Woohoo! Closing now that it's merged (after @anmonteiro rebased it). Very excited to announce this one.

**Pull request closed without merging**